src/IdentityRegistry.ts

```diff
@@ -0,0 +1,137 @@
+import Map from 'dojo-shim/Map';
+import WeakMap from 'dojo-shim/WeakMap';
+import { Handle } from './interfaces';
+
+const noop = () => {};
+
+interface Entry<V> {
+	handle: Handle;
+	value: V;
+}
+
+/**
+ * Registry identities can be strings or symbols. Note that the empty string is allowed.
+ */
+export type Identity = string | symbol;
+
+/**
+ * A registry of values, mapped by identities.
+ */
+export default class IdentityRegistry<V extends Object> {
+	private _entryMap: Map<Identity, Entry<V>>;
+	private _idMap: WeakMap<V, Identity>;
+
+	constructor() {
+		this._entryMap = new Map<Identity, Entry<V>>();
+		this._idMap = new WeakMap<V, Identity>();
+	}
+
+	/**
+	 * Look up a value by its identifier.
+	 *
+	 * Throws if no value has been registered for the given identifier.
+	 *
+	 * @param id The identifier
+	 * @return The value
+	 */
+	get(id: Identity): V {
+		const entry = this._entryMap.get(id);
+		if (!entry) {
+			throw new Error(`Could not find a value for identity '${id.toString()}'`);
+		}
+
+		return entry.value;
+	}
+
+	/**
+	 * Determine whether the value has been registered.
+	 * @param value The value
+	 * @return `true` if the value has been registered, `false` otherwise
+	 */
+	contains(value: V): boolean {
+		return this._idMap.has(value);
+	}
+
+	/**
+	 * Remove from the registry the value for a given identifier.
+	 * @param id The identifier
+	 * @return `true` if the value was removed, `false` otherwise
+	 */
+	delete(id: Identity): boolean {
+		const entry = this._entryMap.get(id);
+		if (!entry) {
+			return false;
+		}
+
+		entry.handle.destroy();
+		return true;
+	}
+
+	/**
+	 * Determine whether a value has been registered for the given identifier.
+	 * @param id The identifier
+	 * @return `true` if a value has been registered, `false` otherwise
+	 */
+	hasId(id: Identity): boolean {
```

Comment:
Why not `has`?  I can guess you might say "well I am not passing the object" but that is similiar to a Map, in that you pass the [key](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/has), not the object.